ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -31,6 +45,245 @@ let tdb_tool = !Xapi_globs.tdb_tool
+  with e -> Error e
+
+let call_wbinfo (args : string list) : (string, exn) result =
+  try
+    (* we trust wbinfo will not print any sensitive info on failure *)
+    let stdout = Helpers.call_script ~log_output:On_failure wb_cmd args in
+    Ok stdout
+  with e -> Error e
+
+module Ldap = struct
+  type user = {
+      upn: string
+    ; account_disabled: bool
+    ; account_expired: bool
+    ; account_locked: bool
+    ; password_expired: bool
+  }
+  [@@deriving rpcty]
+
+  let string_of_user x =
+    Rpcmarshal.marshal user.Rpc.Types.ty x |> Jsonrpc.to_string
+
+  let parse_user stdout : (user, string) result =
+    let module Map = Map.Make (String) in
+    let module P = struct
+      open Angstrom
+
+      let is_space = function ' ' -> true | _ -> false
+
+      let s = take_while is_space
+
+      let ns = take_while @@ fun x -> not (is_space x)
+
+      let is_whitespace = function
+        | ' ' | '\n' | '\t' | '\r' ->
+            true
+        | _ ->
+            false
+
+      let ws = skip_while is_whitespace
+
+      (* example inputs: "key: value\n" or "key: value with spaces\r\n" *)
+      let kvp =
+        let* key = s *> take_while (fun x -> x <> ':') <* char ':' in
+        let* value =
+          s *> take_while (function '\n' | '\r' -> false | _ -> true)
+          <* (end_of_line <|> end_of_input)
+        in
+        return (key, value)
+
+      let header =
+        let* num_replies =
+          choice ~failure_msg:"unexpected header"
+            [string "Got" *> char ' ' *> ns <* char ' ' <* string "replies"]
+        in
+        match num_replies with
+        | "1" ->
+            return ()
+        | _ ->
+            Printf.sprintf "got %s replies" num_replies |> fail
+
+      let p =
+        let* () = ws *> header <* ws in
+        let* l = ws *> many kvp <* ws <* end_of_input in
+        return (l |> List.to_seq |> Map.of_seq)
+
+      let parse (x : string) : (string Map.t, string) result =
+        parse_string ~consume:All p x
+    end in
+    let ldap fmt = fmt |> Printf.kprintf @@ Printf.sprintf "ldap %s" in
+    let* kvps = P.parse stdout <!> ldap "parsing failed '%s'" in
+    let get_string k =
+      match Map.find_opt k kvps with
+      | None ->
+          Error (ldap "missing key '%s'" k)
+      | Some x ->
+          Ok x
+    in
+    let get_int of_string k =
+      let* str = get_string k in
+      try Ok (of_string str)
+      with _ -> Error (ldap "invalid value for key '%s'" k)
+    in
+    let* upn = get_string "userPrincipalName" in
+    let* user_account_control = get_int Int32.of_string "userAccountControl" in
+    let* account_expires = get_int Int64.of_string "accountExpires" in
+    let account_expired =
+      (* see https://docs.microsoft.com/en-us/windows/win32/adschema/a-accountexpires *)
+      let windows_nt_time_to_unix_time x =
+        Int64.sub (Int64.div x 10000000L) 11644473600L
+      in
+      match account_expires with
+      | 0L | 9223372036854775807L ->
```

Comment:
What is this magic number?

Reply:
this number in particular is actually `Int64.max_int`, so I'll replace that

(it's mentioned in that link above)